porting over task guide, mdx bugfixes (#13754)
PrefectHQ/prefect: docs/3.0rc/guides/deferred-tasks.mdx

---
title: Deferred Tasks
description: Examples of using Prefect tasks and the task server.
---

## Why use deferred tasks?

> Prefect tasks are a great way to quickly execute small, discrete units of work. _Deferred_ Prefect tasks run in a background process using a Prefect task server. Deferred tasks to move work out of the foreground of your application and distribute concurrent execution across across multiple processes or machines.

Reviewer: ```suggestion
Prefect tasks are a great way to quickly execute small, discrete units of work. _Deferred_ Prefect tasks run in a background process using a Prefect task server. Use deferred tasks to move work out of the foreground of your application and distribute concurrent execution across across multiple processes or machines.
```

> For example, if you have a web app, you can use tasks to offload processes such as sending emails, processing images, or inserting data into a database.

Reviewer: ```suggestion
For example, if you have a web app, you can use deferred tasks to offload processes such as sending emails, processing images, or inserting data into a database.
```

```

### Calling tasks

You can call a task to run it immediately, or you can defer the task by scheduling it for background execution with `Task.delay`.

**NOTE**: It is also possible to submit tasks to a _task runner_ such as Ray or Dask -- and to defer task execution -- within a workflow, which in Prefect is called a _flow_. However, this document will focus on deferring task execution outside of workflows. For example, by calling `my_task.delay()` within a web application.

However you run a task, Prefect will use your task configuration to manage and control task execution.
The following example shows both methods mentioned earlier, calling a task and using `delay`:

```python
# Import the previously-defined task
from my_tasks import my_background_task

# Run the task immediately
my_background_task("Joaquim")

# Schedule the task for execution outside of this process
my_background_task.delay("Agrajag")
```

For documentation on the features available for tasks, refer to the [Prefect Tasks documentation](https://docs-3.prefect.io/3.0rc/tutorial/tasks).

### Executing deferred tasks with a task server

To run tasks in a separate process or container, start a task server.

> The task server will continually receive deferred tasks to execute from Prefect's API, execute them, and report the results back to the API.

Reviewer: ```suggestion
The task server will continually receive instructions to execute deferred tasks from Prefect's API, execute them, and report the results back to the API.
```

**NOTE:** Task servers only run deferred tasks, not tasks you call directly.

You can run a task server by passing tasks into the `prefect.task_server.serve()` method:

```python tasks.py
from prefect import task
from prefect.task_server import serve

@task
def my_background_task(name: str):
    # Task logic here
    print(f"Hello, {name}!")

if __name__ == "__main__":
    # NOTE: The serve() function accepts multiple tasks. The Task Server 
    # will listen for submitted task runs for all tasks passed in.
    serve(my_background_task)
```

Run this script to start the task server.

> The task server should begin listening for submitted tasks. If tasks were submitted before the task server started, it will begin processing them.

Reviewer: The word "submitted" could potentially be confusing here. We probably reserve that for submission to task runners.

Author: good flag, let me change

---
title: Deferred Tasks
description: Examples of using Prefect tasks and the task server.
---

## Why use deferred tasks?

> Prefect tasks are a great way to quickly execute discrete units of work. _Deferred_ Prefect tasks run in a background process using a Prefect task server. This lets you use deferred tasks to move work out of the foreground of your application and distribute concurrent execution across across multiple processes or machines.

Reviewer: There may be a better word than "small", I would like to at least signal to the user that these are processes that are not meant to take a long time to run.

## Using deferred tasks

Prefect tasks are Python functions that can be run immediately or submitted for background execution, similar to arq tasks. 

You define a task by adding the `@task` decorator to a Python function, after which you can use the `delay` method to run the task in the background.

If you submit the task for background execution, you'll run a task server in a separate process or container to execute the task. This process is similar to how you would run a Celery worker or an arq worker to execute background tasks.

### Defining a task

Add the `@task` decorator to a Python function to define a Prefect task.
Here's an example:

```python
from prefect import task

@task
def my_background_task(name: str):
    # Task logic here
    print(f"Hello, {name}!")
```

### Calling tasks

You can call a task to run it immediately, or you can defer the task by scheduling it for background execution with `Task.delay`.

> **NOTE**: It is also possible to submit tasks to a _task runner_ such as Ray or Dask -- and to defer task execution -- within a workflow, which in Prefect is called a _flow_. However, this document will focus on deferring task execution outside of workflows. For example, by calling `my_task.delay()` within a web application.

Reviewer: This note is very similar to the paragraph beginning on line 12 above. I think we can do without it.

However you run a task, Prefect will use your task configuration to manage and control task execution.
The following example shows both methods mentioned earlier, calling a task and using `delay`:

```python
# Import the previously-defined task
from my_tasks import my_background_task

# Run the task immediately
my_background_task("Joaquim")

# Schedule the task for execution outside of this process
my_background_task.delay("Agrajag")
```

For documentation on the features available for tasks, refer to the [Prefect Tasks documentation](https://docs-3.prefect.io/3.0rc/tutorial/tasks).

### Executing deferred tasks with a task server

> To run tasks in a separate process or container if you schedule them with `delay`, start a task server, similar to how you would run a Celery worker or an arq worker.

Reviewer: ```suggestion
To run tasks in a separate process or container, start a task server.
```

The task server will continually receive deferred tasks to execute from Prefect's API, execute them, and report the results back to the API.

**NOTE:** Task servers only run deferred tasks, not tasks you call directly.

> You can run a task server by passing tasks into the `prefect.task_server.serve()` method, like in the following example *tasks.py* file:

Reviewer: ```suggestion
You can run a task server by passing tasks into the `prefect.task_server.serve()` method:
```

> ```python

Reviewer: ```suggestion
```python tasks.py
```

```

Run this script to start the task server.

The task server should begin listening for submitted tasks. If tasks were submitted before the task server started, it will begin processing them.

You can also use the helper CLI command `prefect task serve` to start a task server.

```bash
prefect task serve my_task.py:my_background_task
```

## Guided exploration of deferred tasks and task servers in Prefect

Below we explore increasingly realistic examples of using deferred tasks and task servers in Prefect.

We'll start by running a Prefect task in the foreground by calling it.

Next we'll start a task server and defer tasks so that they run in the background. We'll see how we can use multiple task servers to run tasks in parallel.

Then we'll create a basic FastAPI application that defers tasks when you hit an endpoint.

Next we'll use Docker in two examples that mimic real use cases.
One example uses a FastAPI server with multiple microservices and simulates a new user signup workflow.
The other example uses a Flask server with [Marvin](https://www.askmarvin.ai/) to ask questions of an LLM from the CLI and get back answers.

The examples build on the ones that come before.

> After setting up your environment, feel free to skip ahead if you're already familiar with the basics.

Reviewer: ```suggestion

```

Use the task run UUID to see the task run in the UI.
The URL will look like this:

`http://127.0.0.1:4200/task-runs/task-run/my_task_run_uuid_goes_here`

Substitute your UUID at the end of the URL.
Note that the UI navigation experience for task runs will be improved soon.

Step 6: You can use multiple task servers to run tasks in parallel.

Start another instance of the task server. In another terminal run:

```bash
python task_server.py
```

Step 7: Submit multiple tasks to the task server.

Modify the `task_submitter.py` file to submit multiple tasks to the task server with different inputs:

```python
from tasks import my_background_task

if __name__ == "__main__":
    my_background_task.delay("Ford")
    my_background_task.delay("Prefect")
    my_background_task.delay("Slartibartfast")
```

Run the file and watch the work get distributed across both task servers!

Step 8: Shut down the task servers with *control* + *c*.

Alright, you're able to submit tasks to multiple Prefect task servers running in the background!
This is cool because we can observe these tasks executing in parallel and very quickly with web sockets - no polling required.

> Check out additional examples in the [deferred tasks github repository](https://github.com/PrefectHQ/prefect-background-task-examples.git).

Reviewer: ```suggestion
See additional examples in the [deferred tasks github repository](https://github.com/PrefectHQ/prefect-background-task-examples.git).
```
